lib/eph/blobLease.ts

```diff
@@ -0,0 +1,308 @@
+
+  partitionId?: string;
+  leaseId?: string;
+  storageAccount: string;
+  blobService: BlobService;
+  containerName: string;
+  blob: string;
+  fullUri: string;
+  hostName: string;
+
+  private _isHeld: boolean = false;
+  private _containerAndBlobExist: boolean = false;
+
+  constructor(hostName: string, storageConnectionString: string, containerName: string, blob: string) {
+    this.hostName = hostName;
+    this.blobService = createBlobService(storageConnectionString);
+    this.storageAccount = (storageConnectionString.match("AccountName=([^;]*);") || [])[1];
+    this.containerName = containerName;
+    this.blob = blob;
+    this.fullUri = `https://${this.storageAccount}.blob.core.windows.net/${containerName}/${blob}`;
+    this._isHeld = false;
+    this._containerAndBlobExist = false;
+    debug("[%s] Full lease path: '%s'.", this.hostName, this.fullUri);
+  }
+
+  /**
+   * Ensures that the container and blob exist.
+   */
+  async ensureContainerAndBlobExist(): Promise<void> {
+    try {
+      if (!this._containerAndBlobExist) {
+        await this._ensureContainerExists();
+        await this._ensureBlobExists();
+        this._containerAndBlobExist = true;
+      }
+    } catch (err) {
+      const msg = `An error occurred while ensuring that the container and blob exists. ` +
+        `It is: \n${JSON.stringify(err)}`;
+      throw new Error(msg);
+    }
+  }
+
+  /**
+   * Returns the best-guess as to whether the lease is still held. May not be accurate if
+   * the lease has expired.
+   * @returns {boolean}
+   */
+  get isHeld(): boolean {
+    return this._isHeld;
+  }
+
+  /**
+   * Since others may manage lease renewal/acquisition, this allows them to tell the lease whether
+   * they believe it is held or not. For instance, if the LeaseManager fails to renew the lease
+   * once, the lease may still be held, but after multiple times, the hold might expire.
+   * The LeaseManager may choose to tell the lease that it has lost the hold before that
+   * has actually occurred.
+   *
+   * The lease is normally pretty good about managing this itself
+   * (on acquire/renew/release success), but for special cases (like the above) this method
+   * might be required.
+   *
+   * @param isItHeld
+   */
+  set isHeld(isItHeld: boolean) {
+    this._isHeld = isItHeld;
+  }
+
+  async acquire(options: BlobService.AcquireLeaseRequestOptions): Promise<BlobLease> {
+    await defaultLock.acquire(Constants.ensureContainerAndBlob, () => { return this.ensureContainerAndBlobExist(); });
+    return await this._acquireLease(options);
+  }
+
+  renew(options: BlobService.AcquireLeaseRequestOptions): Promise<BlobLease> {
+    return new Promise((resolve, reject) => {
+      if (!this.leaseId) {
+        reject(new Error(BlobLease.notHeldError));
+      } else {
+        this.blobService.renewLease(this.containerName, this.blob, this.leaseId, options, (error, result, response) => {
```

Comment:
nit: this is a style thing but i'd just promisify this call and move everything else outside the `new Promise()`